fdbserver/MasterProxyServer.actor.cpp

```diff
@@ -134,40 +144,70 @@ ACTOR Future<Void> getRate(UID myID, Reference<AsyncVar<ServerDBInfo>> db, int64
 	}
 }
 
+struct TransactionRateInfo {
+	double rate;
+	double limit;
+
+	TransactionRateInfo(double rate) : rate(rate), limit(0) {}
+
+	void reset(double elapsed) {
+		limit = std::min(0.0, limit) + rate * elapsed; // Adjust the limit based on the full elapsed interval in order to properly erase a deficit
+		limit = std::min(limit, rate * SERVER_KNOBS->START_TRANSACTION_BATCH_INTERVAL_MAX); // Don't allow the rate to exceed what would be allowed in the maximum batch interval
+		limit = std::min(limit, SERVER_KNOBS->START_TRANSACTION_MAX_TRANSACTIONS_TO_START);
+	}
+
+	bool canStart(int64_t numAlreadyStarted) {
+		return numAlreadyStarted < limit;
+	}
+
+	void updateBudget(int64_t numStarted) {
+		limit -= numStarted;
+	}
+};
+
 ACTOR Future<Void> queueTransactionStartRequests(
-	std::priority_queue< std::pair<GetReadVersionRequest, int64_t>, std::vector< std::pair<GetReadVersionRequest, int64_t> > > *transactionQueue,
-	FutureStream<GetReadVersionRequest> readVersionRequests,
-	PromiseStream<Void> GRVTimer, double *lastGRVTime,
-	double *GRVBatchTime, FutureStream<double> replyTimes,
-	ProxyStats* stats) 
-{
+    Reference<AsyncVar<ServerDBInfo>> db,
+    std::priority_queue<std::pair<GetReadVersionRequest, int64_t>,
+                        std::vector<std::pair<GetReadVersionRequest, int64_t>>>* transactionQueue,
+    FutureStream<GetReadVersionRequest> readVersionRequests, PromiseStream<Void> GRVTimer, double* lastGRVTime,
+    double* GRVBatchTime, FutureStream<double> replyTimes, ProxyStats* stats, TransactionRateInfo* batchRateInfo) {
 	state int64_t counter = 0;
 	loop choose{
 		when(GetReadVersionRequest req = waitNext(readVersionRequests)) {
-			if (req.debugID.present())
+			if (req.debugID.present()) {
 				g_traceBatch.addEvent("TransactionDebug", req.debugID.get().first(), "MasterProxyServer.queueTransactionStartRequests.Before");
+			}
 
-			stats->txnStartIn += req.transactionCount;
-			if (req.priority() >= GetReadVersionRequest::PRIORITY_SYSTEM_IMMEDIATE)
+			if (req.priority() >= GetReadVersionRequest::PRIORITY_SYSTEM_IMMEDIATE) {
 				stats->txnSystemPriorityStartIn += req.transactionCount;
-			else if (req.priority() >= GetReadVersionRequest::PRIORITY_DEFAULT)
+			} else if (req.priority() >= GetReadVersionRequest::PRIORITY_DEFAULT) {
 				stats->txnDefaultPriorityStartIn += req.transactionCount;
-			else
+			} else {
+				// Return error for bath_priority GRV requests
+				int64_t proxiesCount = std::max((int)db->get().client.proxies.size(), 1);
+				if (batchRateInfo->rate <= (1.0 / proxiesCount)) {
+					TEST(true);
+					TraceEvent(SevInfo, "RejectedBatchGRV").detail("CurrentBatchRateLimit", batchRateInfo->limit);
```

Comment:
I suspect that this is going to be pretty verbose, we should consider adding a suppression. Also, I see we're logging the limit but using the rate to determine whether to throttle, was it intention to have that difference?

Reply:
> I suspect that this is going to be pretty verbose, we should consider adding a suppression

Done. 

>  I see we're logging the limit but using the rate to determine whether to throttle, was it intention to have that difference?

That's my bad. Should log `rate` here.